nodes/generators_extended/torus2D.py

```diff
@@ -0,0 +1,273 @@
+#  as published by the Free Software Foundation; either version 2
+#  of the License, or (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, write to the Free Software Foundation,
+#  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# ##### END GPL LICENSE BLOCK #####
+
+import bpy
+from bpy.props import IntProperty, FloatProperty, BoolProperty, EnumProperty
+
+from math import sin, cos, pi
+
+from sverchok.node_tree import SverchCustomTreeNode
+from sverchok.data_structure import updateNode, match_long_repeat
+
+
+def torus_verts(R, r, N1, N2, p, Separate):
+    '''
+        R   : major radius
+        r   : minor radius
+        N1  : major sections - number of RADIAL sections
+        N2  : minor sections - number of CIRCULAR sections
+        p   : radial section phase
+    '''
+    listVerts = []
+
+    # angle increments (cached outside of the loop for performance)
+    da = 2*pi/N1
+
+    for n1 in range(N1):
+        theta = n1 * da + p     # radial section angle
+        sin_theta = sin(theta)  # caching
+        cos_theta = cos(theta)  # caching
+
+        loopVerts = []
+        s = 2 / (N2-1)  # caching
+        for n2 in range(N2):
+            rr = R + (n2*s - 1) * r
+            x = rr * cos_theta
+            y = rr * sin_theta
+
+            # append vertex to loop
+            loopVerts.append([x, y, 0.0])
+
+        if Separate:
+            listVerts.append(loopVerts)
+        else:
+            listVerts.extend(loopVerts)
+
+    return listVerts
+
+
+def torus_edges(N1, N2):
+    '''
+        N1 : major sections - number of RADIAL sections
+        N2 : minor sections - number of CIRCULAR sections
+    '''
+    listEdges = []
+
+    # radial EDGES
+    for n1 in range(N1):
+        for n2 in range(N2-1):
+            listEdges.append([N2*n1 + n2, N2*n1 + n2+1])
+
+    # circular EDGES
+    for n1 in range(N1-1):
+        for n2 in range(N2):
+            listEdges.append([N2*n1 + n2, N2*(n1+1) + n2])
+    for n2 in range(N2):
+        listEdges.append([N2*(N1-1) + n2, N2*0 + n2])
+
+    return listEdges
+
+
+def torus_polygons(N1, N2):
+    '''
+        N1 : major sections - number of RADIAL sections
+        N2 : minor sections - number of CIRCULAR sections
+    '''
+    listPolys = []
+    for n1 in range(N1-1):
+        for n2 in range(N2-1):
+            listPolys.append(
+                [N2*n1 + n2, N2*(n1+1) + n2, N2*(n1+1) + n2+1, N2*n1 + n2+1])
+
+    for n2 in range(N2-1):
+        listPolys.append([N2*(N1-1) + n2, N2*0 + n2, N2*0 + n2+1, N2*(N1-1) + n2+1])
+    listPolys.append([N2*(N1-1) + N2-1, N2*0 + N2-1, N2*0 + 0, N2*(N1-1) + 0])
+
```

Comment:
I can clean these out. I wrote it this way to better understand the code as it shows that that is actually the edges at the zero index.  These no-ops are in the regular torus code btw. :) you missed those.